File: Swift/Playgrounds/Advent of Code 2023.swiftpm/Solutions/Day19.swift
import Foundation

class Day19: AoCSolution {
    override init() {
        super.init()
        self.day = 19
        self.name = "Aplenty"
        self.emptyLinesIndicateMultipleInputs = false
    }
        
    override func solve(_ input: AoCInput) -> AoCResult {
        super.solve(input)
        
        let (workflowList, metalBits) = parseInput(input.textLines)
        var workflows = Dictionary<String, Workflow>()
        for w in workflowList {
            workflows[w.name] = w
        }
        
        let part1 = solvePartOne(workflows, shapes: metalBits)
        let part2 = solvePartTwo(workflows)
        
        return AoCResult(part1: "\(part1)", part2: "\(part2)")
    }
    
    func solvePartOne(_ workflows: Dictionary<String,Workflow>, 
                      shapes bits:[WeirdMetalShape]) -> Int {
        var accepted = [WeirdMetalShape]()
        
        for bit in bits {
            var name = "in"
            while true {
                let target = workflows[name]!.eval(shape: bit)
                if target == "A" {
                    accepted.append(bit)
                    break
                }
                else if target == "R" {break}
                name = target
            }
        }
        
        var ratingSum = 0
        for bit in accepted {
            ratingSum += bit.rating
        }
        
        return ratingSum
    }
    
    func solvePartTwo(_ workflows: Dictionary<String, Workflow>) -> Int {
        var acceptNodes = [WFTreeNode]()
        // Builds tree recursively. We only really need the A leaf nodes
        let _ = WFTreeNode(workflow: workflows["in"]!,
                              parent: nil,
                              allWorkflows: workflows,
                              allAcceptNodes: &acceptNodes)
        var sum = 0
        for node in acceptNodes {
            // Traverse to the root, picking up the conditions that got here
            var ptr = node
            var negateNext = false
            var conditions = [WorkflowCondition]()
            while true {
                var cond = ptr.condition
                if cond != nil {
                    if negateNext {
                        cond = cond!.negated()
                        negateNext = false
                    }
                    conditions.append(cond!)
                    //print("\(ptr.name): \(cond!)")
                }
                let isLeftChild = ptr.parent == nil || (ptr.name == ptr.parent!.childL!.name)
                if !isLeftChild {
                    // This was on the "false" branch so the parent's condition was opposite
                    negateNext = true
                }
                if ptr.parent == nil {break}
                ptr = ptr.parent!
            }
            var summaryByLetter = ["x": 4000,"m": 4000,"a": 4000,"s": 4000]
            for key in ["x","m","a","s"] {
                summaryByLetter[key] = WorkflowCondition.combine(conditions.filter({$0.key == key}))
                //print("\(key): \(summaryByLetter[key]!), ", terminator: "")
            }
            //print()
            let product = summaryByLetter.values.reduce(1, *)
            //print(product)
            sum += product
        }
        return sum
    }
    
    func parseInput(_ input: [String]) -> ([Workflow], [WeirdMetalShape]) {
        var workflows = [Workflow]()
        var i = 0
        var line = input[i]
        while !line.isEmpty {
            workflows.append(Workflow(line))
            i += 1
            line = input[i]
       }
        
        var bits = [WeirdMetalShape]()
        for j in i+1..<input.count {
            line = input[j]
            bits.append(WeirdMetalShape(line))
        }
        
        return (workflows, bits)
    }
}

struct WorkflowCondition: CustomDebugStringConvertible, Hashable {
    let key: String
    let relation: String
    let value: Int
    
    init(_ str: String) {
        let re = #/([xmas])([<>])(\d+)/#
        if let match = str.firstMatch(of: re) {
            key = String(match.1)
            relation = String(match.2)
            value = Int(String(match.3))!
        }
        else {
            key = ""
            relation = ""
            value = 0
        }
    }
    
    // Once you implement a non-default init, you have to define the default yourself
    init(key: String, relation: String, value: Int) {
        self.key = key
        self.relation = relation
        self.value = value
    }
    
    func eval(value input: Int) -> Bool {
        switch relation {
        case "<": return input < value
        case "<=": return input <= value
        case ">": return input > value
        case ">=": return input >= value
        default: return true // avoiding compilation error
        }
    }
    
    func negated() -> WorkflowCondition {
        let opposite: String
        switch relation {
        case "<": opposite = ">="
        case "<=": opposite = ">"
        case ">": opposite = "<="
        case ">=": opposite = "<"
        default: opposite = "ERROR" // avoiding compilation error
        }
        return WorkflowCondition(key: self.key, relation: opposite, value: self.value)
    }
    
    static func combine(_ conditions: [WorkflowCondition]) -> Int {
        if conditions.isEmpty { return 4000 }
        var low = 0 // greater than 
        var high = 4000 // less than or equal to
        
        for condition in conditions {
            if condition.relation == "<" {
                high = min(high, condition.value - 1)
            }
            else if condition.relation == "<=" {
                high = min(high, condition.value)
            }
            else if condition.relation == ">" {
                low = max(low, condition.value)
            }
            else if condition.relation == ">=" {
                low = max(low, condition.value - 1)
            }
        }
        
        return high - low
    }
    
    var debugDescription: String {
        return "\(key)\(relation)\(value)"
    }
}

struct WorkflowRule: CustomDebugStringConvertible, Hashable {
    let condition: WorkflowCondition?
    let target: String
    
    init(_ str: String) {
        let parts = str.split(separator: ":")
        if parts.count == 1 {
            condition = nil
            target = str
        }
        else {
            condition = WorkflowCondition(String(parts.first!))
            target = String(parts.last!)
        }
    }
    
    func eval(shape bit:WeirdMetalShape) -> String? {
        if let condition {
            if condition.eval(value: bit.data[condition.key]!) {
                return target
            }
            return nil
        }
        return target
    }
    
    var debugDescription: String {
        return "\(String(describing: condition)):\(target)"
    }
}

struct Workflow: CustomDebugStringConvertible, Hashable {
    let name: String 
    let rules: [WorkflowRule]
    
    // Once you implement a non-default init, you have to define the default yourself
    init(name: String, rules: [WorkflowRule]) {
        self.name = name
        self.rules = rules
    }
    
    init(_ str: String) {
        // ex. px{a<2006:qkq,m>2090:A,rfg}
        let re = #/^(\w+){(.+)}$/#
        if let match = str.firstMatch(of: re) {
            name = String(match.1)
            var ruleTemp = [WorkflowRule]()
            let parts = String(match.2).split(separator: ",")
            for part in parts {
                let rule = WorkflowRule(String(part))
                ruleTemp.append(rule)
            }
            rules = ruleTemp
        }
        else {
            name = "ERROR"
            rules = [WorkflowRule]()
        }
    }
    
    func eval(shape bit:WeirdMetalShape) -> String {
        for rule in rules {
            if let target = rule.eval(shape: bit) {
                return target
            }
        }
        return "ERROR" // Last rule always is unconditional. Will never get here.
    }
    
    func split() -> (WorkflowRule, Workflow?) {
        assert(!rules.isEmpty)
        let firstRule = rules.first!
        var remainder: Workflow? = nil
        if rules.count > 1 {
            remainder = Workflow(name: self.name + "*",
                                   rules: Array(rules[1...]))
        }
        return (firstRule, remainder)
    }
    
    var debugDescription: String {
        return "\(name){\(rules)}"
    }
}

struct WeirdMetalShape: CustomDebugStringConvertible {
    let data: Dictionary<String,Int>
    var x: Int { return data["x"]! }
    var m: Int { return data["m"]! }
    var a: Int { return data["a"]! }
    var s: Int { return data["s"]! }
    
    init(_ str: String) {
        // ex. {x=787,m=2655,a=1222,s=2876}
        let re = #/x=(\d+),m=(\d+),a=(\d+),s=(\d+)/#
        if let match = str.firstMatch(of: re) {
            let nums = [match.1, match.2, match.3, match.4].map {Int(String($0))!}
            data = ["x": nums[0], "m": nums[1], "a": nums[2], "s": nums[3]]
        }
        else {
            data = ["x": 0, "m": 0, "a": 0, "s": 0]
        }
    }
    
    var rating: Int {
        return data.values.reduce(0, +)
    }
    
    var debugDescription: String {
        return "x:\(x) m:\(m) a:\(a) s:\(s)"
    }
}

class WFTreeNode {
    var parent: WFTreeNode? = nil
    var childL: WFTreeNode? = nil
    var childR: WFTreeNode? = nil
    let name: String
    let condition: WorkflowCondition?
    
    init(workflow wf: Workflow, 
         parent p:WFTreeNode?, 
         allWorkflows: Dictionary<String,Workflow>,
         allAcceptNodes: inout [WFTreeNode]) {
        parent = p
        name = wf.name
        let (rule, remainder) = wf.split()
        self.condition = rule.condition
        if rule.target == "A" || rule.target == "R" {
            childL = WFTreeNode(name: rule.target, parent: self)
            if rule.target == "A" { allAcceptNodes.append(childL!) }
        }
        else {
            childL = WFTreeNode(workflow: allWorkflows[rule.target]!,
                                parent: self,
                                allWorkflows: allWorkflows,
                                allAcceptNodes: &allAcceptNodes)
        }
        if let remainder {
            childR = WFTreeNode(workflow: remainder, 
                                parent: self, 
                                allWorkflows: allWorkflows,
                                allAcceptNodes: &allAcceptNodes)
        }
    }
    
    init(name: String, parent p: WFTreeNode?) {
        self.parent = p
        self.name = name
        self.condition = nil
    }
}

File: Perl/day19.pl
#!/usr/bin/env perl
use v5.42;
use feature 'class';
no warnings qw( experimental::class );

our $directory;
BEGIN { use Cwd; $directory = cwd; }
use lib $directory . '/lib';

use feature 'signatures';
use Data::Printer;
use List::Util qw( sum product );

use AOC::Util;

my $DAY = sprintf("%02d", 19);
my $INPUT = $ARGV[0];
my $INPUT_FILE = 'day' . $DAY . '_' . $INPUT . '.txt';

my @inputs = read_grouped_input("../Input/$INPUT_FILE");

say "Advent of Code 2023, Day 19: Aplenty";

my %workflows = parse_workflows(@{$inputs[0]});
my @parts = parse_parts(@{$inputs[1]});

solve_part_one();
solve_part_two();

exit( 0 );

sub solve_part_one() {
	my @accepted_ratings = ();
	for my $part (@parts) {
		my $w_name = 'in';

		do {
			my $workflow = $workflows{$w_name};
			$w_name = $workflow->evaluate($part);
		} until $w_name eq 'A' || $w_name eq 'R';

		push(@accepted_ratings, $part->rating()) if $w_name eq 'A';
	}

	my $total = sum @accepted_ratings;

	say "Part One: the sum of ratings of the accepted parts is $total.";
}

sub solve_part_two() {
	my @all_accept_nodes = ();
	my $tree;
	$tree = WFTreeNode->build_tree($workflows{'in'}, $tree, \@all_accept_nodes);

	my $sum = 0;
	for my $node (@all_accept_nodes) {
		my $ptr = $node;
		my $negate_next = 0;
		my @conditions = ();

		while (defined $ptr) {
			my $cond = $ptr->condition();
			if (defined $cond) {
				if ($negate_next) {
					$cond = $cond->negated();
					$negate_next = 0;
				}
				push(@conditions, $cond);
# 				say $ptr->name() . ': ' . $cond->to_str();
			}

			my $is_left_child = !(defined $ptr->parent() ) ||
								 $ptr->name() eq $ptr->parent()->child_L()->name();
			if (!$is_left_child) {
				$negate_next = 1;
			}

			$ptr = $ptr->parent();
		}

		my %summary_by_letter = ("x" => 4000, "m" => 4000, "a" => 4000, "s" => 4000);
		for my $key ('x', 'm', 'a', 's') {
			$summary_by_letter{$key} = WorkflowCondition->combine(grep {$_->key() eq $key} @conditions);
		}
# 		say join(', ', map {$_ . ': ' . $summary_by_letter{$_}} ('x', 'm', 'a', 's'));

		my $p = product(values %summary_by_letter);
# 		say $p;
		$sum += $p;
	}
	say "Part Two: the total number of possible combinations is $sum.";
}

sub parse_workflows(@input) {
	my @list = map { Workflow->new(src => $_) } @input;
	my %result = ();
	for my $w (@list) {
		$result{$w->name} = $w;
	}
	return %result;
}

sub parse_parts(@input) {
	my @list = map { WeirdMetalShape->new(src => $_) } @input;
	return @list;
}


class WeirdMetalShape {
	field $src :param :reader;
	field $x :reader;
	field $m :reader;
	field $a :reader;
	field $s :reader;

	ADJUST {
		$src =~ m/\{x=(\d+),m=(\d+),a=(\d+),s=(\d+)\}/;
		$x = $1; $m = $2; $a = $3; $s = $4;
	}

	method rating() {
		return $x + $m + $a + $s;
	}

	method value($key) {
		return $x if $key eq 'x';
		return $m if $key eq 'm';
		return $a if $key eq 'a';
		return $s if $key eq 's';
		return 0;
	}
}

class WorkflowCondition {
	field $src :param :reader;
	field $key :reader;
	field $relation :reader;
	field $value :reader;

	ADJUST {
		if ($src =~ m/([xmas])([<>=]+)(\d+)/) {
			$key = $1;
			$relation = $2;
			$value = $3;
		}
		else {
			$key = "";
			$relation = "";
			$value = 0;
		}
	}

	method evaluate($input) {
		if ($relation eq '<') { return $input < $value }
		if ($relation eq '<=') { return $input <= $value }
		if ($relation eq '>') { return $input > $value }
		if ($relation eq '>=') { return $input >= $value }
		return 1;
	}

	method negated() {
		my $opposite;
		if ($relation eq '<') { $opposite = '>=' }
		if ($relation eq '<=') { $opposite = '>' }
		if ($relation eq '>') { $opposite = '<=' }
		if ($relation eq '>=') { $opposite = '<' }
		return WorkflowCondition->new(src => "$key$opposite$value");
	}

	method to_str() {
		return "$key$relation$value";
	}

	sub combine($cls, @conditions) {
		return 4000 if scalar(@conditions == 0);
		use List::Util qw( min max );
		my $low = 0;
		my $high = 4000;

		for my $cond (@conditions) {
			if ($cond->relation() eq '<') {
				$high = min($high, $cond->value() - 1);
			}
			elsif ($cond->relation() eq '<=') {
				$high = min($high, $cond->value());
			}
			elsif ($cond->relation() eq '>') {
				$low = max($low, $cond->value());
			}
			elsif ($cond->relation() eq '>=') {
				$low = max($low, $cond->value() - 1);
			}
		}
		return $high - $low;
	}
}

class WorkflowRule {
	field $src :param :reader;
	field $condition :reader;
	field $target :reader;

	ADJUST {
		my @parts = split(/:/, $src);
		if (scalar(@parts) == 1) {
			$target = $src;
		}
		else {
			$condition = WorkflowCondition->new(src => $parts[0]);
			$target = $parts[1];
		}
	}

	method to_str() {
		return $condition->to_str() . ":" . $target if defined $condition;
		return $target;
	}

	method evaluate($wms) {
		if (defined $condition) {
			return $target if $condition->evaluate($wms->value($condition->key()));
			return '';
		}
		return $target;
	}

}

class Workflow {
	field $src :param :reader;
	field $name :reader;
	field $rules = [];

	ADJUST {
		$src =~ m/(\w+)\{(.+)\}/;
		$name = $1;
		my @parts = split(/,/, $2);
		my @rules = ();
		for my $rule_src (@parts) {
			push(@rules, WorkflowRule->new(src => $rule_src));
		}
		$rules = \@rules;
	}

	method evaluate($wms) {
		for my $rule (@{$rules}) {
			my $result = $rule->evaluate($wms);
			return $result if $result ne '';
		}
	}

	method split() {
		my @r = @{$rules};
		my $first_rule = $r[0];
		my $remainder;
		if (scalar @r > 1) {
			my @other_rules = @r[1..$#r];
			my $n = $name . 'X';
			my @other_rules_str = map {$_->to_str()} @other_rules;
			my $rule_str = join(',', @other_rules_str);
			$remainder = Workflow->new(src => $n . "{$rule_str}");
		}

		return ($first_rule, $remainder);
	}
}

class WFTreeNode {
	field $parent :reader :writer;
	field $child_L :reader :writer;
	field $child_R :reader :writer;
	field $name :reader :writer;
	field $condition :reader :writer;

	sub build_tree($cls, $workflow, $parent, $inout_a_nodes) {
		my $node = WFTreeNode->new();
		$node->set_parent($parent);
		$node->set_name($workflow->name());
		my ($rule, $remainder) = $workflow->split();
		$node->set_condition($rule->condition());
		if ($rule->target() eq 'A' || $rule->target() eq 'R') {
			my $leaf = WFTreeNode->new();
			$leaf->set_name($rule->target());
			$leaf->set_parent($node);
			$node->set_child_L($leaf);
			push(@{$inout_a_nodes}, $leaf) if $rule->target() eq 'A';
		}
		else {
			$node->set_child_L(WFTreeNode->build_tree($workflows{$rule->target()}, $node, $inout_a_nodes));
		}
		if (defined $remainder) {
			$node->set_child_R(WFTreeNode->build_tree($remainder, $node, $inout_a_nodes));
		}
		return $node;
	}
}
